### PR TITLE
PS-5305: Fix log encryption issues reported by ASAN (5.7)

### PR DESCRIPTION
**Pull request by dutow**

### CODE REVIEW

File: sql/mysqld.cc [hunk 1]
```diff
@@ -8743,7 +8743,7 @@ static int fix_paths(void)
   my_realpath(mysql_home,mysql_home,MYF(0));
   /* Ensure that mysql_home ends in FN_LIBCHAR */
   pos=strend(mysql_home);
-  if (pos[-1] != FN_LIBCHAR)
+  if (pos==mysql_home || pos[-1] != FN_LIBCHAR)
```
Comment (laurynas-biveinis) [inline sql/mysqld.cc]: What is being fixed here? How is it related to encryption?

File: storage/innobase/log/log0log.cc [hunk 2]
```diff
@@ -1432,16 +1432,26 @@ log_enable_encryption_if_set()
 					} else {
 						memcpy(key, rkey2, ENCRYPTION_KEY_LEN);
 					}
+					if(rkey2 != NULL) {
```
Comment (laurynas-biveinis) [inline storage/innobase/log/log0log.cc]: Space after if

File: storage/innobase/log/log0log.cc [hunk 3]
```diff
@@ -1432,16 +1432,26 @@ log_enable_encryption_if_set()
...
+					}
 				}
+			} else {
+				memcpy(key, rkey, ENCRYPTION_KEY_LEN);
 			}
 
 			if (encryption_enabled) {
 				ut_ad(redo_key_type && strcmp(redo_key_type, "AES") == 0);
+			}
+			if(redo_key_type != NULL) {
```
Comment (laurynas-biveinis) [inline storage/innobase/log/log0log.cc]: Space after if

File: storage/innobase/log/log0log.cc [hunk 4]
```diff
@@ -1432,16 +1432,26 @@ log_enable_encryption_if_set()
...
 				my_free(redo_key_type);
 			}
+			if(rkey != NULL) {
```
Comment (laurynas-biveinis) [inline storage/innobase/log/log0log.cc]: Space after if

File: storage/innobase/log/log0log.cc [hunk 5]
```diff
@@ -1517,6 +1527,16 @@ log_enable_encryption_if_set()
 			if (!log_write_encryption(NULL, NULL)) {
 				ib::error() << "Can't write redo log encryption information.";
 			}
+			if(rkey != NULL) {
```
Comment (laurynas-biveinis) [inline storage/innobase/log/log0log.cc]: Spaces after ifs

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]
Please provide descriptive commit message

File: sql/mysqld.cc (hunk 1 above)
Comment (dutow) [reply sql/mysqld.cc]: If mysql_home is empty, pos[-1] is a buffer underflow - which happens in one of the encryption testcases. I'll mention this in the commit message.

Comment (laurynas-biveinis) [reply sql/mysqld.cc]: So a new upstream bug?

**Comment by dutow** [comment]
Updated.

File: storage/innobase/log/log0log.cc [hunk 6]
```diff
@@ -1432,16 +1432,26 @@ log_enable_encryption_if_set()
 					} else {
 						memcpy(key, rkey2, ENCRYPTION_KEY_LEN);
 					}
+					if (rkey2 != NULL) {
+						my_free(rkey2);
```
Comment (laurynas-biveinis) [inline storage/innobase/log/log0log.cc:1436]: For 8.0, consider rewriting with map_helpers.h:unique_ptr_my_free

**Review by laurynas-biveinis — APPROVED** [review]
Since the mysqld.cc bug was fixed in 8.0, the normal thing to do here would be to cherry-pick its ce1fd349764018574ba6c7dc000dbc73279222bd in a separate commit, but now OK to merge as-is

**Comment by dutow** [comment]
https://ps2.cd.percona.com/job/percona-server-5.7-param/24/#showFailuresLink

**Pull request merged**